### PR TITLE
chore: Adjust authentication policy

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
## Changes
* Addressed comments from https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2937
* Fixed failing tests caused by this change
* Changed and added multiple tests connected to auth policies
* Adjusted a few parts of the SDK implementation (using enums where possible, added a few missing parts, etc.)

## TODO
* Mention in https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2880 that the SDK for Auth Policies is ready

### CODE REVIEW

File: pkg/sdk/testint/authentication_policies_gen_integration_test.go [hunk 1]
```diff
@@ -14,331 +16,274 @@ import (
...
+			authenticationPolicies, err := client.AuthenticationPolicies.Show(ctx, sdk.NewShowAuthenticationPolicyRequest().
+				WithLike(sdk.Like{Pattern: sdk.String("test_auth_policy_2_%")}).
+				WithIn(sdk.In{Schema: id.SchemaId()}))
+			require.NoError(t, err)
+			assert.Len(t, authenticationPolicies, 1)
+		})
 
-		err = client.AuthenticationPolicies.Drop(ctx, sdk.NewDropAuthenticationPolicyRequest(id))
-		require.NoError(t, err)
+		// TODO(ticket number): starts_with doesn't work (returns all)
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/authentication_policies_gen_integration_test.go:228]: ticket number?

File: pkg/sdk/testint/authentication_policies_gen_integration_test.go [hunk 2]
```diff
@@ -14,331 +16,274 @@ import (
...
+				WithLimit(sdk.LimitFrom{Rows: sdk.Int(1)}).
+				WithIn(sdk.In{Schema: id.SchemaId()}))
+			require.NoError(t, err)
+			assert.Len(t, authenticationPolicies, 1)
+		})
 
-		assert.LessOrEqual(t, 2, len(returnedAuthenticationPolicies))
-		assert.Contains(t, returnedAuthenticationPolicies, *authenticationPolicy1)
-		assert.Contains(t, returnedAuthenticationPolicies, *authenticationPolicy2)
+		// TODO(ticket number): limit from doesn't work (should return 0 elements because alphabetically test_auth_policyzzz is last in the output)
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/authentication_policies_gen_integration_test.go:263]: ticket number?

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

**Review by sfc-gh-jmichalak — DISMISSED** [review]
+1 for ticket numbers, otherwise ✅

File: pkg/sdk/users.go [hunk 3]
```diff
@@ -390,15 +390,15 @@ func (opts *RemoveDelegatedAuthorization) validate() error {
...
 	ObjectProperties     *UserAlterObjectProperties `ddl:"keyword"`
 	ObjectParameters     *UserObjectParameters      `ddl:"keyword"`
 	SessionParameters    *SessionParameters         `ddl:"keyword"`
 }
 
 func (opts *UserSet) validate() error {
-	if !exactlyOneValueSet(opts.PasswordPolicy, opts.SessionPolicy, opts.AuthenticationPolicy, opts.ObjectProperties, opts.ObjectParameters, opts.SessionParameters) {
-		return errExactlyOneOf("UserSet", "PasswordPolicy", "SessionPolicy", "AuthenticationPolicy", "ObjectProperties", "ObjectParameters", "SessionParameters")
+	if !anyValueSet(opts.PasswordPolicy, opts.SessionPolicy, opts.AuthenticationPolicy, opts.ObjectProperties, opts.ObjectParameters, opts.SessionParameters) {
+		return errAtLeastOneOf("UserSet", "PasswordPolicy", "SessionPolicy", "AuthenticationPolicy", "ObjectProperties", "ObjectParameters", "SessionParameters")
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/users.go:401]: This condition is true; at least one field has to be set, but you're right. The validation seems to be not complete. I added more cases and tests that allow only valid combinations.

File: pkg/sdk/testint/authentication_policies_gen_integration_test.go (hunk 1 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/authentication_policies_gen_integration_test.go:228]: my bad, added

File: pkg/sdk/testint/authentication_policies_gen_integration_test.go (hunk 2 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/authentication_policies_gen_integration_test.go:263]: Added

File: pkg/sdk/users.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/users.go:401]: Hmmm, okay, so please take a look at the `alter: set and unset parameters` test too. It was splitting the unset into two (and I imagine that the reason being this `keyword` <> `list` diff). Additionally, the unset is also split inside the resource, so please adjust the implementation there too.

File: pkg/sdk/users.go [hunk 4]
```diff
@@ -426,8 +432,14 @@ type UserUnset struct {
 
 func (opts *UserUnset) validate() error {
 	// TODO [SNOW-1645875]: change validations with policies
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/users.go:434]: this TODO is still valid after these changes?

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/sdk/users.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/users.go:401]: Done

File: pkg/sdk/users.go (hunk 4 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/users.go:434]: It is; it's about adding support of { auth | password | session } policies on user resource.

Comment (sfc-gh-asawicki) [reply pkg/sdk/users.go:434]: But you have already added the validations for them (and the TODO is about changing the validations with policies). Is something moar needed here?

**Review by sfc-gh-asawicki — APPROVED** [review]

File: pkg/sdk/users.go (hunk 4 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/users.go:434]: From the ticket description, adding them to the user schema (?)

Comment (sfc-gh-asawicki) [reply pkg/sdk/users.go:434]: Comment was left, but it seems that there are no more validations needed. Will be removed.

**Pull request merged**